File: Advanced Programming Methods/Toy Language/src/Model/MyStack.java
package Model;

import java.util.*;

//Execution Stack
public class MyStack<T> implements IStack<T> {
    private Stack<T> stack;
    private int nr;

    public MyStack(){
        this.stack = new Stack<>();
    }
    @Override
    public void push(T el) {
        stack.push(el);
        ++nr;
    }

    @Override
    public T pop() throws MyException{
        if(isEmpty())
            throw new MyException("Empty stack");
        --nr;
        return stack.pop();
    }

    @Override
    public int length() {
        return nr;
    }

    @Override
    public boolean isEmpty() {
        return stack.isEmpty();
    }

    public String toString(){
        String res = "ExeStack:";
        Collections.reverse(stack);
        ArrayList<T> list = new ArrayList<>(stack);
        for(Object el:list){
            res = res.concat(System.lineSeparator());
            res = res.concat(el.toString());
        }
        Collections.reverse(stack);
        return res;
    }
}

File: Advanced Programming Methods/Toy Language C#/A2/Model/Data Structures/MyStack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace A2.Model.Data_Structures
{
    //Execution Stack
    public class MyStack<T> : IStack<T> {
        private Stack<T> stack;
        private int nr;

        public MyStack()
        {
            stack = new Stack<T>();
        }

        public void push(T el)
        {
            stack.Push(el);
            ++nr;
        }
        
        public T pop()
        {
            if(isEmpty())
                throw new MyException("Empty stack");
            --nr;
            return stack.Pop();
        }
        public int length()
        {
            return nr;
        }
    
        public Boolean isEmpty()
        {
            return stack.Count() == 0;
        }

        public override string ToString()
        {
            String res = "ExeStack:";
            List<T> list = new List<T>(stack);
            //list.Reverse();
            foreach (Object el in list)
            {
                res += Environment.NewLine;
                res += String.Concat(el.ToString());
            }

            return res;
        }
    }

}
